### ISSUE TITLE
Qwen implementation has no direct-align

### ISSUE THREAD

**Issue by avramdj** [Tencent-Hunyuan/SRPO#42]
Your qwen approach has no inverse branch. Is this intentional? Any new findings?

**Comment by xwwshen** [comment]
Thanks，the inverse branch is an engineering trick that, via gradient accumulation, injects negative optimization directions during training; this can further reduce reward-hacking and push toward extreme aesthetic results. However, it tends to hurt training efficiency, so we omitted it in this release to better support more general objectives like ID consistency and lower artifact rates.

**Comment by ustc-yaojinchen** [comment]
Thanks for the detailed explanation!I have a quick question for clarification:
Isn’t artifact exactly one type of reward hacking?If our goal is to get lower artifact rates, should we use the inverse branch or not use it?
Looking forward to your advice!

**Comment by xwwshen** [comment]
> Thanks for the detailed explanation!I have a quick question for clarification: Isn’t artifact exactly one type of reward hacking?If our goal is to get lower artifact rates, should we use the inverse branch or not use it? Looking forward to your advice!

Thank you for the questions. By "artifacts" I mean things like extra fingers or distorted/unnatural poses. We tried training a reward model specifically to detect these cases. Reward hacking usually shows up as greasy/overly shiny textures and color shifts. If you’re concerned about hacking, I can suggest enabling inversion or trying our discount_pos parameter — you can set the early discount to a negative value to penalize these high-frequency rewards, for example: -0.1 0.2.

**Comment by ustc-yaojinchen** [comment]
Thanks a lot for your detailed reply! It really clarifies the differences between artifacts and reward hacking for me.
I have a couple of follow-up questions:
Does the original Direct-Align approach include the inverse branch?
If I want to reproduce Direct-Align (including the inverse branch if it’s part of it), could I modify the code in the way shown in below?
Even a brief guidance on the modification direction would be greatly appreciated! Thanks again for your help.


```
def SRPO_train()
            ################

            with torch.amp.autocast('cuda'):
                # if  inversion ==1:
                #     outputs = (reward_model.SRP_cfg([pos_caption],batch_caption,image,discount[mid_timestep])) 
                # else:
                #     outputs = (reward_model.SRP_cfg(batch_caption,[pos_caption],image,discount_inversion[mid_timestep]))
                
                # Revise the logic of reward_model.SRP_cfg such that when the input prompt is None, the relevant calculations involving this prompt are excluded from the reward computation.
                if  inversion ==1:
                    outputs = (reward_model.SRP_cfg([pos_caption],None,image,discount[mid_timestep])) 
                else:
                    outputs = (reward_model.SRP_cfg(None,[pos_caption],image,discount_inversion[mid_timestep]))

            # Follow ReFL set a threshold
            loss = F.relu(-outputs+0.7)/gradient_accumulation_steps
            loss = loss.mean()
            loss.backward()
    return loss
```

**Comment by xwwshen** [comment]
> Thanks a lot for your detailed reply! It really clarifies the differences between artifacts and reward hacking for me. I have a couple of follow-up questions: Does the original Direct-Align approach include the inverse branch? If I want to reproduce Direct-Align (including the inverse branch if it’s part of it), could I modify the code in the way shown in below? Even a brief guidance on the modification direction would be greatly appreciated! Thanks again for your help.
> 
> ```
> def SRPO_train()
>             ################
> 
>             with torch.amp.autocast('cuda'):
>                 # if  inversion ==1:
>                 #     outputs = (reward_model.SRP_cfg([pos_caption],batch_caption,image,discount[mid_timestep])) 
>                 # else:
>                 #     outputs = (reward_model.SRP_cfg(batch_caption,[pos_caption],image,discount_inversion[mid_timestep]))
>                 
>                 # Revise the logic of reward_model.SRP_cfg such that when the input prompt is None, the relevant calculations involving this prompt are excluded from the reward computation.
>                 if  inversion ==1:
>                     outputs = (reward_model.SRP_cfg([pos_caption],None,image,discount[mid_timestep])) 
>                 else:
>                     outputs = (reward_model.SRP_cfg(None,[pos_caption],image,discount_inversion[mid_timestep]))
> 
>             # Follow ReFL set a threshold
>             loss = F.relu(-outputs+0.7)/gradient_accumulation_steps
>             loss = loss.mean()
>             loss.backward()
>     return loss
> ```

You should remove the inversion=0 part. Your reward code logic is correct, but we don't yet support None as an input, so you can just delete the corresponding calculation.

**Comment by ustc-yaojinchen** [comment]
Thank you for your reply.
